src/Core/Services/UsernameGenerationService.cs

```diff
@@ -0,0 +1,250 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Bit.Core.Abstractions;
+using Bit.Core.Enums;
+using Bit.Core.Models.Domain;
+using Bit.Core.Utilities;
+
+namespace Bit.Core.Services
+{
+    public class UsernameGenerationService : IUsernameGenerationService
+    {
+        private const string DefaultGenerated = "-";
+        private readonly ICryptoService _cryptoService;
+        private readonly IApiService _apiService;
+        private readonly IStateService _stateService;
+        private UsernameGenerationOptions _defaultOptions = new UsernameGenerationOptions(true);
+        private UsernameGenerationOptions _optionsCache;
+
+        public UsernameGenerationService(
+            ICryptoService cryptoService,
+            IApiService apiService,
+            IStateService stateService
+            )
+        {
+            _cryptoService = cryptoService;
+            _apiService = apiService;
+            _stateService = stateService;
+        }
+
+        public async Task<string> GenerateUsernameAsync(UsernameGenerationOptions options)
+        {
+            switch (options.Type)
+            {
+                case UsernameType.PlusAddressedEmail:
+                    return await GeneratePlusAddressedEmailAsync(options);
+                case UsernameType.CatchAllEmail:
+                    return await GenerateCatchAllAsync(options);
+                case UsernameType.ForwardedEmailAlias:
+                    return await GenerateForwardedEmailAliasAsync(options);
+                case UsernameType.RandomWord:
+                    return await GenerateRandomWordAsync(options);
+                default:
+                    return string.Empty;
```

Comment:
I'd use the `ILogger` here and log an `Error` saying that there is an options type that is not implemented to generate a username, and then return `string.Empty`. So we have visibility on AppCenter just in case